smartconfig/constructors.py

```diff
@@ -0,0 +1,47 @@
+from typing import Any
+
+import yaml
+
+from smartconfig.exceptions import InvalidOperation
+from smartconfig._registry import get_attribute, get_node
+
+_used_constructors = set()
+
+
+class AttributeReference:
+    """
+    A descriptor referencing an attribute in a config entry.
+
+    Args:
+        path: A dot-delimited path to the attribute being referenced.
+    """
+
+    def __init__(self, path: str) -> None:
+        if path in _used_constructors:
+            raise InvalidOperation("Cannot point a REF constructor to another one.")
```

Comment:
Fixed by walking the whole reference path in 81f3728.

Reply:
Why did you go for this approach instead of storing the path? Is disallowing references to paths which are also references not sufficient to prevent circular references? Looping to check each constructor for circular dependencies seems more complex and slower.

Reply:
> Is disallowing references to paths which are also references not sufficient to prevent circular references?
It would be, although we can't point a ref to another ref which would be pretty lame.

Reply:
Okay, that's fine then. I wish it didn't need to have `follow_descriptors` forwarded like 4 times, but I don't have any better ideas. It's an implementation detail so we can come back and address it later if we want.